### ISSUE TITLE
Grouping: Strange behavior when grouping by date.

### ISSUE THREAD

**Issue by tlsim** [angular-ui/ng-grid#3629]
http://plnkr.co/edit/6SqMr5anUrHBzAfiPj50?p=preview

Expected to get 2 groups, each with 3 child rows. Instead get 3 instances of each group row with a single duplicated child row.

Possibly related: sortingAlgorithm is called with date objects as expected, but then is called with group cell string content - e.g. "Wed Apr 01 2015 00:00:00 GMT+0100 (GMT Standard Time) (1)".

Should cellFilter work with the group rows, or do I need a custom cell template?

**Comment by AgDude** [comment]
Thanks for the report. There are two separate issues here.

1. We are using equality (```!==```) to identify groups, so two date objects will never be equal.

2. It looks like we aren't applying the cellFilter to group header rows.

**Comment by PaulL1** [comment]
We should be applying cellFilter to group headers - they're treated the same as any other row.  But sometimes filters don't work if we've added counts to the cell value - that then breaks some filters.  The tutorial has an example of how to get around that.

**Comment by AgDude** [comment]
@PaulL1 is there any reason not to apply the cellFilter in our default render method?

**Comment by AgDude** [comment]
I was thinking a bit more about this. What about an option to group after "cellFilter" is applied? There are two places I can see it will be an issue for me.

1. A date column where I am interested in the day of the week, not the actual date.
2. A number column where I have applied a cellFilter to round the value

It also occurred to me that I don't think we support ```customAggrgationFinalizerFn``` on grouped columns. That should be an easy fix, but I'm unlikely to get to it until next week sometime.

**Comment by PaulL1** [comment]
My problem with cellFilter has always been that it's defined in a binding format - so it's `currency | translate | myCustomThing`

If we want to run it in code we'd have to unstring it, and in theory we can call $filter('myCustomThing')(value) I think, but I'm not sure that's reliable because all those filters can return promises.  Where I wanted to use filters in things like menus I specified that users had to provide a function that ran synchronously - so for example with angular-translate you use $translate.instant.  

Bottom line: I'm not sure we can easily or reliably run the filters in code to be able to group after applying it.  

We could create a new setting, groupFilter, which has a function that we can call before grouping, ideally it'd be in the same format I did for some other features.....exporter perhaps.....so that people don't have lots of slightly different formats to fill in in different parts of the application.

**Comment by AgDude** [comment]
I was thinking ```$filter('myCustomThing')(value)```. But I think you're right that if we have chained filters it isn't that simple.  I also agree we want things to be consistent. 

The other issue which we are going to have to solve at some point is sorting after the cellFilter is applied. So maybe what we need is a util of some sort that parses the filters properly and can be used by all of the features. I don't mind taking that on, but I can't promise when.

**Comment by PaulL1** [comment]
Yeah, every time I start down that path I also start thinking about entirely removing the cell bindings, applying the filters in code, and then only updating the cell when the reference changes.  Then I think about how to do that in custom templates.  Then I remember that watchers are only a problem because of scrolling (mostly), so maybe I'm better to work out how to not run watchers during scrolling, which I have a plan to do but haven't done.  Then I think maybe we're better to focus on shipping 3.0.

**Comment by AgDude** [comment]
So clearly this issue exposes some problems that aren't quick fixes. But dates are going to be the most common by far, so let's check for dates in grouping processor, and call ``getTime``` on them before comparison.

Better handling of cellFilters throughout would be a great improvement, but I wouldn't holdup 3.0 for it.
